picky-asn1-x509/src/name.rs

```diff
@@ -341,8 +346,63 @@ impl<'de> de::Deserialize<'de> for GeneralName {
 
 // OtherName ::= SEQUENCE {
 //      type-id    OBJECT IDENTIFIER,
-//      value      [0] EXPLICIT ANY DEFINED BY type-id }
-//pub struct OtherName { ... }
+//      value      [0] EXPLICIT ANY DEFINED BY type-id}
+#[derive(Debug, PartialEq, Clone)]
+pub struct OtherName {
+    pub type_id: ObjectIdentifierAsn1,
+    pub value: ExplicitContextTag0<picky_asn1_der::Asn1RawDer>,
+}
+
+impl ser::Serialize for OtherName {
+    fn serialize<S>(&self, serializer: S) -> Result<<S as ser::Serializer>::Ok, <S as ser::Serializer>::Error>
+    where
+        S: ser::Serializer,
+    {
+        let mut buff = Vec::new();
+        buff.extend_from_slice(
+            &picky_asn1_der::to_vec(&self.type_id).map_err(|e| ser::Error::custom(format!("{:?}", e)))?,
+        );
+        buff.extend_from_slice(
+            &picky_asn1_der::to_vec(&self.value).map_err(|e| ser::Error::custom(format!("{:?}", e)))?,
+        );
+
+        ExplicitContextTag0::from(Asn1RawDer(buff)).serialize(serializer)
+    }
+}
```

Comment:
Can't this be derived?

Reply:
This one can't be. `SEQUENCE` tag is replaced with `ExplicitContextTag0`.
![image](https://user-images.githubusercontent.com/55661041/144898897-e3be4ee2-33c0-4eb3-b3a4-f044f7b65141.png)
Making `Serialize` derived would force the `SEQUENCE` tag to be present. Is there a proper solution?

Reply:
I understand the `SEQUENCE` tag must indeed be present as per RFC 4043:
![image](https://user-images.githubusercontent.com/3809077/145093114-9d1b1ac2-c855-486d-a0e1-ad256ce439f2.png)
As such, the following (the one in your message) is appropriate encoding:
![image](https://user-images.githubusercontent.com/3809077/145093191-860c5ebc-9573-4781-963e-0893eab0da12.png)
(not sure if you meant something else?)

Reply:
Oooh, sorry, I understand! You meant as opposed to this:
![image](https://user-images.githubusercontent.com/3809077/145093988-f740d6ba-7774-43a6-8560-b9b04097f64c.png)
Technically, that's how it should be encoded standalone:
```
// OtherName ::= SEQUENCE {
//      type-id    OBJECT IDENTIFIER,
//      value      [0] EXPLICIT ANY DEFINED BY type-id}
```

But you're right in that the `GeneralName` choice use an _IMPLICIT_ [0] tag to wrap that. This means that when serializing a `GeneralName` specifically we have to write the [0] tag instead of the `SEQUENCE` tag.

I suggest you just derive `Serialize` on `OtherName`, and modify the manual `Serialize` implementation of `GeneralName` in the "other name" match arm like so:
```rust
        let mut raw_der = picky_asn1_der::to_vec(&self.0).unwrap_or_else(|_| vec![0]);
        raw_der[0] = Tag::context_specific_constructed(0).inner();
        picky_asn1_der::Asn1RawDer(raw_der).serialize(serializer)
```
(still a workaround because we have no proper way to work with that as of now)

Since I did the modification locally to make sure it worked, I just committed it directly: https://github.com/Devolutions/picky-rs/pull/112/commits/bddc0df9cec61717c6205c83aa4f63eff661adf3

Reply:
Thank you!

Reply:
@CBenoit  
Do you think we could add support for such situations in the future? 
Maybe, you should open an issue to not forget about it. Someone could even fix this :slightly_smiling_face:

Reply:
Yes, I would like a proper solution to this, but it requires a deep refactoring:
- Updating the serde `Serializer` and `Deserializer` implementations (not `Serialize`, `Deserialize` but the actual serializer code). This is quite involved.
- Adding a ASN.1 value type similar to `serde_json::Value` for loose deserialization (which should replace the `RawDerAsn1` use cases).
- Updating the code using the old approach
- I'm not sure how to fix everything yet and probably need a few experimentation

This will be required when we add BER support (for one, we can't depend on the `RawDerAsn1` in this case), so I plan on doing that on this occasion :slightly_smiling_face: (or sooner if I get the chance, though Devolutions Gateway is taking my time right now)